test/e2e/framework/framework.go

```diff
@@ -84,6 +84,8 @@ type Framework struct {
 	NamespaceDeletionTimeout time.Duration
 	SkipPrivilegedPSPBinding bool // Whether to skip creating a binding to the privileged PSP in the test namespace
 
+	SecondaryNamespace *v1.Namespace
```

Comment:
Why is this needed inside the framework? If the CSI driver deployment code is the only place where this is needed, then it may be better handle the second namespace there.

Reply:
Having secondary namespace being tracked here has several benefits. 
- The code in framework.go ensures any registered namespace is deleted even if somehow it is skipped from storage testsuite.
- It ensures proper reporting and dumping of events in case there is an error deleting the namespace etc.

Reply:
Why can't the same be done in the storage test suite? Tests should have the same access to cleanup and logging as the framework itself.

Reply:
may be. But I do not see any inherent advantage of doing this in storage test suite vs doing it here. I also forgot to mention - if there are projects that use k8s e2e as a library then having the namespace exposed in framework.go allows them to customize it. Such as apply extra permissions etc.

Reply:
I'm asking because there has been a recent push slim down the API of the framework, including the removal of code that was only used by some tests. For example, podlog was originally added to the framework "because it might be useful to others" and then moved under test/e2e/storage because in practice it was only used by storage tests.

Reply:
okay, but in this case IMO it makes sense to have this in framework.go because moving this code to storage suite will result in unnecessary duplication.

Reply:
The cleanup enhancements in the framework belong there, but I don't see the need for `CreateSecondaryNamespace()` and `f.SecondaryNamespace` in the framework.

What will be duplicated when moving that into `test/e2e/storage`? `CreateSecondaryNamespace` only uses exported functionality of the framework, so it can be implemented outside of the framework without copying code.

Reply:
yes the function itself has to be copied and then following bits of code that runs in `AfterEach` of `framework.go` needs to be copied:

```
if err := f.ClientSet.CoreV1().Namespaces().Delete(context.TODO(), ns.Name, metav1.DeleteOptions{}); err != nil {
					if !apierrors.IsNotFound(err) {
						nsDeletionErrors[ns.Name] = err

						// Dump namespace if we are unable to delete the namespace and the dump was not already performed.
						if !ginkgo.CurrentGinkgoTestDescription().Failed && TestContext.DumpLogsOnFailure {
							DumpAllNamespaceInfo(f.ClientSet, ns.Name)
						}
					} else {
						Logf("Namespace %v was already deleted", ns.Name)
					}
				}
```

So as namespace info could be dumped. I have not looked very closely though there could be more stuff and that is what worries me.

Reply:
Perhaps the framework should have a generic API for deleting additional namespaces in a guaranteed order and a function for creating them? That looks like a cleaner approach to me than `f.SecondaryNamespace` - what if a test needs three or more namespaces...

Reply:
hmm okay but I am unsure about how that API will look like. @pohly do you think it is a blocker issue for merging this PR?

Another thing about creating that generic API is - if it exists at framework initialization time then it has to be triggered via testsuite whereas creating secondary namespace for now is part of CSI driver. So basically, a storage test that runs with in-tree EBS and CSI gce-pd driver behave differently and framework initialization happens at the top of e2e test file.

Reply:
> hmm okay but I am unsure about how that API will look like. @pohly do you think it is a blocker issue for merging this PR?

Given that moving the framework into staging is a goal for 1.19 (https://github.com/kubernetes/enhancements/issues/1746) and that this implies stabilizing the API, I'm worried about this rather ad-hoc API change. But I'll let @alejandrox1 and @oomichi decide this.

Reply:
okay, so I moved most of namespace creation code to storage/utils.go . The namespace cleanup should still happen because framework.go tracks these namespaces in `namespacesToDelete` variable.